### PR TITLE
Register wallet data files component only if user has created wallets

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/33757
Resolves https://github.com/brave/brave-browser/issues/30940

This PR does following things:
  - For normal users without wallet: wallet data files won't be registered.
  - For first time wallet users, wallet data files component will be registered
      when Create/ResotreWallet and ImportFromExternalWallet APIs are called, the
      finish callback of these APIs won't be run until wallet data files component
      is ready and content is parsed into BlockchainRegistry.
  - For existing wallet users, the behavior is unchanged, wallet data files
      component will continue to be registered and parsed upon startup.
  - Remove json sanitization on JSON files from brave wallet data
      files component served via component updater which is considered trusted data.

Note that if the installation of component or parsing failed, we won't retry and will unblock the APIs without error in this case.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Screenshots:
Android - Create Wallet|Android - Restore Wallet
--|--
https://github.com/brave/brave-core/assets/24739341/3fe5936c-8f51-450f-96e8-9a92d2de4af5|https://github.com/brave/brave-core/assets/24739341/76d2a094-f5e4-4710-a52f-47b3fd12bc80

### Create New Wallet

https://github.com/brave/brave-core/assets/40611140/585bd4b6-8c6d-4b9e-9850-3941c54152d3

### Import From Existing Wallet

https://github.com/brave/brave-core/assets/40611140/b5a032e2-4dad-405a-9e5d-32ed63543f87

### Restore From Seed Phrase

https://github.com/brave/brave-core/assets/40611140/145d7b1e-2fe6-4c37-b0ff-b6b01eb7ae79

## Test Plan:
1. Start with a new user data dir folder (clean state), `npm start -- --user_data_dir_name=PR20741` or run the executable with `--user-data-dir=some_path`. (Should see a welcome page when started)
2. Go to brave://components, Brave Wallet data files should not present.
3. Go to brave://wallet, create a new wallet
4. Visit brave://wallet/crypto/portfolio/add-asset#available-assets, should see tokens, and Brave Wallet data files should present in brave://components
5. Start with another new user data dir folder
6. Go to brave://components, Brave Wallet data files should not present.
7. Go to brave://wallet, import from an existing wallet to restore from seed phrase.
8. Visit brave://wallet/crypto/portfolio/add-asset#available-assets, should see tokens, and Brave Wallet data files should present in brave://components

Also, do the similar steps for testing `import from existing wallet` like above video shows.

### CODE REVIEW

File: components/brave_wallet/browser/blockchain_registry.cc [hunk 1]
```diff
@@ -308,4 +496,64 @@ bool BlockchainRegistry::IsOfacAddress(const std::string& address) {
...
+          [](const base::FilePath& absolute_install_dir) {
+            ParseCoingeckoIdsMapAndUpdateRegistry(absolute_install_dir);
+            ParseTokenListAndUpdateRegistry(absolute_install_dir);
+            ParseChainListAndUpdateRegistry(absolute_install_dir);
+            ParseDappListsAndUpdateRegistry(absolute_install_dir);
+            ParseOnRampListsAndUpdateRegistry(absolute_install_dir);
+            ParseOfacAddressesListsAndUpdateRegistry(absolute_install_dir);
+          },
+          absolute_install_dir),
+      std::move(callback));
```
Comment (supermassive) [inline components/brave_wallet/browser/blockchain_registry.cc:535]: 1) We don't need two calls on MayBlock runner. Just one which resolves path and parses jsons.
2) So instead `sequenced_task_runner_` just `base::ThreadPool::PostTaskAndReplyWithResult(FROM_HERE, {base::MayBlock(), ...`
3) That blocking operation should produce an 'internal state' of BlockchainRegistry:
  ```
    CoingeckoIdsMap coingecko_ids_map_;
    TokenListMap token_list_map_;
    etc...
  ```
  and return it as `...ReplyWithResult`
4) `BlockchainRegistry::GetInstance()` object is now(and actually was) accessed from different threads and should be fixed by (3). Also makes sense to add DCHECK_CALLED_ON_VALID_SEQUENCE for instance of BlockchainRegistry::GetInstance() internal static.

File: components/brave_wallet/browser/keyring_service.cc [hunk 2]
```diff
@@ -1101,15 +1102,32 @@ void KeyringService::CreateWallet(const std::string& mnemonic,
...
   ResetAllAccountInfosCache();
-  std::move(callback).Run(mnemonic);
+
+  WalletDataFilesInstaller::GetInstance()
+      .MaybeRegisterWalletDataFilesComponentOnDemand(base::BindOnce(
+          [](const std::string& mnemonic, CreateWalletCallback callback) {
+            std::move(callback).Run(mnemonic);
+          },
+          mnemonic, std::move(callback)));
 }
```
Comment (supermassive) [inline components/brave_wallet/browser/keyring_service.cc:1112]: How much time will it take to setup component?
Do we need some spinner with label like "Wallet is being initialized" or so?

Comment (yrliou) [reply components/brave_wallet/browser/keyring_service.cc:1112]: Yes, we have spinner, desktop commit https://github.com/brave/brave-core/pull/20741/commits/6d77b80c9d5525b81e6ff3adf336a0a5fb618d0d.
Android too in this PR, iOS UI PR will be covered by https://github.com/brave/brave-ios/issues/8345.

File: components/brave_wallet/browser/blockchain_registry.cc (hunk 1 above)
Comment (yrliou) [reply components/brave_wallet/browser/blockchain_registry.cc]: Fixed by https://github.com/brave/brave-core/pull/20741/commits/f1e4487746ffded61f315d30f8555d3b8f95aff7.

**Review by supermassive — APPROVED** [review]
wallet core lgtm!

**Review by StephenHeaps — APPROVED** [review]
iOS++. 
Hit https://github.com/brave/brave-browser/issues/34095 but also hit it on `master` so seems unrelated to this PR.

File: components/brave_wallet/browser/brave_wallet_utils.cc [hunk 3]
```diff
@@ -1217,6 +1217,10 @@ void UpdateLastUnlockPref(PrefService* prefs) {
   prefs->SetTime(kBraveWalletLastUnlockTime, base::Time::Now());
 }
 
+bool HasCreatedWallets(PrefService* prefs) {
+  return !prefs->GetTime(kBraveWalletLastUnlockTime).is_null();
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/brave_wallet_utils.cc:1221]: Fix by https://github.com/brave/brave-core/pull/20741/commits/4d16c00cfe9216eaa84188c5a2f4d040f3f67c6f.

**Review by SergeyZhukovsky — APPROVED** [review]
android part ++

**Comment by yrliou** [comment]
Remove needs-security-review flag and those assignees due to base::Unretained(this) which is removed now.

File: components/brave_wallet_ui/page/screens/onboarding/create-password/onboarding-create-password.tsx [hunk 4]
```diff
@@ -51,6 +56,9 @@ export const OnboardingCreatePassword = (
   const isWalletCreated = useSelector(
     ({ wallet }: { wallet: WalletState }) => wallet.isWalletCreated
   )
+  const isCreatingWallet = useSelector(
+    ({ page }: { page: PageState }) => page.isCreatingWallet
```
Comment (josheleonard) [inline components/brave_wallet_ui/page/screens/onboarding/create-password/onboarding-create-password.tsx:60]: let's create a reusable selector function for this

File: components/brave_wallet_ui/page/screens/onboarding/onboarding.routes.tsx [hunk 5]
```diff
@@ -44,6 +44,9 @@ export const OnboardingRoutes = () => {
   const termsAcknowledged = useSelector(
     ({ page }: { page: PageState }) => page.walletTermsAcknowledged
   )
+  const isCreatingWallet = useSelector(
+    ({ page }: { page: PageState }) => page.isCreatingWallet
```
Comment (josheleonard) [inline components/brave_wallet_ui/page/screens/onboarding/onboarding.routes.tsx:48]: let's use a reusable selector function here

File: components/brave_wallet_ui/page/screens/onboarding/restore-from-recovery-phrase/restore-from-recovery-phrase.tsx [hunk 6]
```diff
@@ -96,8 +93,8 @@ export const OnboardingRestoreFromRecoveryPhrase = ({
   const importWalletAttempts = useSelector(
     ({ page }: { page: PageState }) => page.importWalletAttempts
   )
-  const isWalletCreated = useSelector(
-    ({ wallet }: { wallet: WalletState }) => wallet.isWalletCreated
+  const isCreatingWallet = useSelector(
+    ({ page }: { page: PageState }) => page.isCreatingWallet
```
Comment (josheleonard) [inline components/brave_wallet_ui/page/screens/onboarding/restore-from-recovery-phrase/restore-from-recovery-phrase.tsx:97]: let's use a reusable selector

**Review by josheleonard — CHANGES_REQUESTED** [review]

File: components/brave_wallet_ui/page/screens/onboarding/create-password/onboarding-create-password.tsx (hunk 4 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/onboarding/create-password/onboarding-create-password.tsx]: Done!

File: components/brave_wallet_ui/page/screens/onboarding/onboarding.routes.tsx (hunk 5 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/onboarding/onboarding.routes.tsx]: Done!

File: components/brave_wallet_ui/page/screens/onboarding/restore-from-recovery-phrase/restore-from-recovery-phrase.tsx (hunk 6 above)
Comment (Douglashdaniel) [reply components/brave_wallet_ui/page/screens/onboarding/restore-from-recovery-phrase/restore-from-recovery-phrase.tsx]: Done!

File: components/brave_wallet_ui/page/screens/onboarding/restore-from-recovery-phrase/restore-from-recovery-phrase.tsx [hunk 7]
```diff
@@ -52,6 +55,7 @@ import { CenteredPageLayout } from '../../../../components/desktop/centered-page
 import { StepsNavigation } from '../../../../components/desktop/steps-navigation/steps-navigation'
 import { RecoveryInput } from './recovery-input'
 import { Checkbox } from '../../../../components/shared/checkbox/checkbox'
+import { CreatingWallet } from '../creating_wallet/creating_wallet'
```
Comment (josheleonard) [inline components/brave_wallet_ui/page/screens/onboarding/restore-from-recovery-phrase/restore-from-recovery-phrase.tsx:58]: nit: not a style

**Review by josheleonard — APPROVED** [review]

File: browser/ui/android/strings/android_brave_strings.grd [hunk 8]
```diff
@@ -2085,6 +2085,15 @@ Are you sure you want to do this?
...
       Verified password did not match
     </message>
+    <message name="IDS_CREATING_WALLET_ONBOARDING_TEXT" desc="Text displayed along with a spinner during Brave Wallet creation or restoring.">
+      Creating Wallet...
+    </message>
+    <message name="IDS_YOUR_WALLET_IS_RESTORING_PAGE_TITLE" desc="Text for a spinner page title during Brave Wallet restoring. It is not displayed anywhere" translateable="false">
+      Your wallet is restoring...
+    </message>
+    <message name="IDS_YOUR_WALLET_IS_CREATING_PAGE_TITLE" desc="Text for a spinner page title during Brave Wallet creating. It is not displayed anywhere" translateable="false">
+      Your wallet is creating...
```
Comment (mkarolin) [inline browser/ui/android/strings/android_brave_strings.grd:2095]: Suggestions: `Your wallet is being created...` or `Creating your wallet...`

**Review by mkarolin — APPROVED** [review]
`strings`++

**Review by mkarolin — APPROVED** [review]
`chromium_src`++

File: browser/ui/android/strings/android_brave_strings.grd (hunk 8 above)
Comment (yrliou) [reply browser/ui/android/strings/android_brave_strings.grd]: Changed to `Your wallet is being created...`, thanks.

**Pull request merged**